### ISSUE TITLE
yearfirst bug fixed?

### ISSUE THREAD

**Issue by jreback** [dateutil/dateutil#217]
xref https://github.com/pydata/pandas/issues/12611

I guess this was a bug in < 2.5?

```
In [1]: import dateutil.parser

In [2]: p = dateutil.parser.parser()

In [3]: p.parse(StringIO('090131'), dayfirst=False, yearfirst=False)
Out[3]: datetime.datetime(2009, 1, 31, 0, 0)

In [4]: dateutil.__version__
Out[4]: '2.4.1'
```

```
In [1]: import dateutil.parser
In [3]: p = dateutil.parser.parser()

In [4]: p.parse(StringIO('090131'), dayfirst=False, yearfirst=False)
Out[4]: datetime.datetime(2031, 9, 1, 0, 0)

In [5]: p.parse(StringIO('090131'), dayfirst=False, yearfirst=True)
Out[5]: datetime.datetime(2009, 1, 31, 0, 0)

In [5]: import dateutil

In [6]: dateutil.__version__
Out[6]: '2.5.0'
```

**Comment by pganssle** [comment]
According to `git bisect`, this was fixed in [this commit](https://github.com/dateutil/dateutil/commit/fc6962540e2), which I believe was part of #130. I'll add some tests to make sure it stays fixed.

This is at least partially issue #81.

**Comment by jreback** [comment]
@pganssle great thanks! didn't even realize that we had this edge case until some tests broke :)

anyhow, #214 is related here.

**Comment by jreback** [comment]
@pganssle see that 2.5.3 came out...... here's my new matrix

```
        # OK
        # 2.5.1 10-11-12   [dayfirst=0, yearfirst=0] -> 2012-10-11 00:00:00
        # 2.5.2 10-11-12   [dayfirst=0, yearfirst=1] -> 2012-10-11 00:00:00
        # 2.5.3 10-11-12   [dayfirst=0, yearfirst=0] -> 2012-10-11 00:00:00

        # OK
        # 2.5.1 10-11-12   [dayfirst=0, yearfirst=1] -> 2010-11-12 00:00:00
        # 2.5.2 10-11-12   [dayfirst=0, yearfirst=1] -> 2010-11-12 00:00:00
        # 2.5.3 10-11-12   [dayfirst=0, yearfirst=1] -> 2010-11-12 00:00:00

        # bug fix in 2.5.2
        # 2.5.1 10-11-12   [dayfirst=1, yearfirst=1] -> 2010-11-12 00:00:00
        # 2.5.2 10-11-12   [dayfirst=1, yearfirst=1] -> 2010-12-11 00:00:00
        # 2.5.3 10-11-12   [dayfirst=1, yearfirst=1] -> 2010-12-11 00:00:00

        # OK
        # 2.5.1 10-11-12   [dayfirst=1, yearfirst=0] -> 2012-11-10 00:00:00
        # 2.5.2 10-11-12   [dayfirst=1, yearfirst=0] -> 2012-11-10 00:00:00
        # 2.5.3 10-11-12   [dayfirst=1, yearfirst=0] -> 2012-11-10 00:00:00

        # OK
        # 2.5.1 20/12/21   [dayfirst=0, yearfirst=0] -> 2021-12-20 00:00:00
        # 2.5.2 20/12/21   [dayfirst=0, yearfirst=0] -> 2021-12-20 00:00:00
        # 2.5.3 20/12/21   [dayfirst=0, yearfirst=0] -> 2021-12-20 00:00:00

        # OK
        # 2.5.1 20/12/21   [dayfirst=0, yearfirst=1] -> 2020-12-21 00:00:00
        # 2.5.2 20/12/21   [dayfirst=0, yearfirst=1] -> 2020-12-21 00:00:00
        # 2.5.3 20/12/21   [dayfirst=0, yearfirst=1] -> 2020-12-21 00:00:00

        # revert of bug in 2.5.2
        # 2.5.1 20/12/21   [dayfirst=1, yearfirst=1] -> 2020-12-21 00:00:00
        # 2.5.2 20/12/21   [dayfirst=1, yearfirst=1] -> month must be in 1..12
        # 2.5.3 20/12/21   [dayfirst=1, yearfirst=1] -> 2020-12-21 00:00:00

        # OK
        # 2.5.1 20/12/21   [dayfirst=1, yearfirst=0] -> 2021-12-20 00:00:00
        # 2.5.2 20/12/21   [dayfirst=1, yearfirst=0] -> 2021-12-20 00:00:00
        # 2.5.3 20/12/21   [dayfirst=1, yearfirst=0] -> 2021-12-20 00:00:00
```

**Comment by pganssle** [comment]
@jreback Looks good to me.
